src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5570,36 +5511,40 @@ private static Task InternalWhenAll(Task[] tasks)
         // which involves several allocations, with this logic:
         //      return new WhenAllPromise(tasksCopy);
         // which saves a couple of allocations and enables debugger notification specialization.
-        //
-        // Used in InternalWhenAll(Task[])
         private sealed class WhenAllPromise : Task, ITaskCompletionAction
         {
-            /// <summary>
-            /// Stores all of the constituent tasks.  Tasks clear themselves out of this
-            /// array as they complete, but only if they don't have their wait notification bit set.
-            /// </summary>
-            private readonly Task?[] m_tasks;
```

Comment:
Removing this has a debugging downside.  Previously in a debugger you could look at a Task.WhenAll task and see all of the tasks that comprise it, which have already completed and which haven't (e.g. just by examining the object in the watch window).  There's no longer a back-reference now, making that much harder (you'd need to scour the heap looking for object references).

Reply:
I've added a field for the IEnumerable<Task>, its content should be easy to examine for most IEnumerable implementations.

Reply:
> I've added a field for the IEnumerable

This is unfortunately a potential problem in and of itself.  Imagine a call like:
```C#
await Task.WhenAll(data.Select(() => Task.Run(...));
```
Storing the enumerable such that it might be enumerated again (even in the debugger) would cause a whole new set of tasks to be launched.  It also means that it's not necessarily the same set that contributed to this WhenAll.

Reply:
Isn't it expected by the user with the message "Expanding the Results View will enumerate the IEnumerable" from the debugger? If the implementation of the IEnumerable is a collection the user will know that expanding the result is safe.

Reply:
Even if it's expected, it's not helpful for debugging: in such a case it wouldn't be the same tasks being waited on here.  (And then the moment you do that, you've also launched additional work that could easily corrupt the rest of the diagnostic session.)

Reply:
Could you confirm that with a code like that:
```csharp
await Task.WhenAll(data.Select(() => Task.Run(...).ToArray());
```
expanding the results view will always give the same tasks?

Reply:
Assuming you meant:
```C#
data.Select(... => Task.Run(...)).ToArray()
```
i.e. creating an array from the enumerable, then yes, you'll always get the same tasks instances, since you'll be viewing an array.

Reply:
Then (I'm not sure of the general rule but) if the IEnumerable is a collection, the field is helpful for debugging. Only in the case of an enumeration (not sure how to call that either), displaying the IEnumerable could have side effects. But can't we rely on the fact that developers are aware of those side effects whenever they see an IEnumerable?